src/Nethermind/Nethermind.Consensus/Processing/ParallelProcessing/MultiVersionMemory.cs

```diff
@@ -0,0 +1,194 @@
+// SPDX-FileCopyrightText: 2025 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Nethermind.Core.Collections;
+using Nethermind.Core.Extensions;
+
+namespace Nethermind.Consensus.Processing.ParallelProcessing;
+
+public class MultiVersionMemory<TLocation, TLogger>(ushort txCount, ParallelTrace<TLogger> parallelTrace) where TLogger : struct, IIsTracing where TLocation : notnull
+{
+    private readonly record struct Value(ushort Incarnation, byte[] Bytes);
+    private static readonly Value Estimate = new(ushort.MaxValue, []);
+
+    private readonly ConcurrentDictionary<TLocation, Value>[] _data =
+        Enumerable.Range(0, txCount)
+        .Select(_ => new ConcurrentDictionary<TLocation, Value>())
+        .ToArray();
+
+    private readonly HashSet<TLocation>?[] _lastWrittenLocations = new HashSet<TLocation>?[txCount];
+    private readonly HashSet<Read<TLocation>>?[] _lastReads = new HashSet<Read<TLocation>>[txCount];
+
+    private void ApplyWriteSet(Version version, Dictionary<TLocation, byte[]> writeSet)
+    {
+        ConcurrentDictionary<TLocation, Value> txData = _data[version.TxIndex];
+        foreach (KeyValuePair<TLocation, byte[]> write in writeSet)
+        {
+            txData[write.Key] = new(version.Incarnation, write.Value);
+        }
+    }
+
+    private bool UpdateWrittenLocations(int txIndex, Dictionary<TLocation, byte[]> writeSet)
+    {
+        ConcurrentDictionary<TLocation, Value> txData = _data[txIndex];
+        ref HashSet<TLocation>? lastWritten = ref _lastWrittenLocations[txIndex];
+        lastWritten ??= new HashSet<TLocation>();
+
+        if (lastWritten.Count != 0)
+        {
+            using ArrayPoolList<TLocation> toRemove = new(lastWritten.Count);
```

Comment:
I think I do understand that we remove what was not written this round. I don't understand why.